cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/PullOperation.java

```diff
@@ -0,0 +1,146 @@
+ * {@link io.cdap.cdap.internal.app.services.ApplicationLifecycleService} while running in
+ * app-fabric and through deploy apis when running in task workers.
+ */
+public class PullOperation implements LongRunningOperation<PullOperationRequest> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PullOperation.class);
+
+  private final InMemorySourceControlOperationRunner scmOpRunner;
+  private final SourceControlHelper helper;
+
+  private final Set<ApplicationId> deployed;
+
+  @Inject
+  PullOperation(InMemorySourceControlOperationRunner scmOpRunner, SourceControlHelper helper) {
+    this(scmOpRunner, helper, Clock.systemUTC());
+  }
+
+  @VisibleForTesting
+  private PullOperation(InMemorySourceControlOperationRunner scmOpRunner,
+      SourceControlHelper helper, Clock clock) {
+    this.scmOpRunner = scmOpRunner;
+    this.helper = helper;
+    this.deployed = new HashSet<>();
+  }
+
+  @Override
+  public Type getRequestType() {
+    return PullOperationRequest.class;
+  }
+
+  @Override
+  public ListenableFuture<OperationError> run(PullOperationRequest request,
+      Consumer<OperationMeta> updateMetadata) throws Exception {
+    OperationError error = null;
+    MultiPullAppOperationRequest pullReq = new MultiPullAppOperationRequest(
+        request.getRepoConfig(),
+        request.getNamespace(),
+        request.getApps().stream().map(AppToPull::getApplication).collect(Collectors.toSet())
+    );
+
+    AtomicReference<ApplicationId> appTobeDeployed = new AtomicReference<>();
+
+    try {
+      // pull and deploy applications one at a time
+      scmOpRunner.pull(pullReq, response -> {
+        String versionId = RunIds.generate().getId();
+        appTobeDeployed.set(new ApplicationId(request.getNamespace().getEntityName(),
+            response.getApplicationName(), versionId));
+        helper.deployAppVersion(
+            appTobeDeployed.get(), response
+        );
+        deployed.add(appTobeDeployed.get());
+        updateMetadata.accept(getOperationMeta());
+      });
+    } catch (Exception e) {
+      error = new OperationError(
+          "Failed to deploy application",
+          appTobeDeployed.get() != null ? ImmutableList.of(
+              new OperationResourceScopedError(appTobeDeployed.get().toString(), e.getMessage()))
+              : Collections.emptyList()
+      );
+    }
+
+    // only mark latest if no error happened
+    if (error == null) {
+      try {
+        // all deployed versions are marked latest atomically
+        helper.markAppVersionsLatest(deployed);
+      } catch (Exception e) {
+        error = new OperationError(
+            "Failed to mark applications latest",
+            Collections.emptySet()
+        );
+      }
+    }
+
+    if (error != null) {
+      // If error happened we try to delete all deployed versions
+      for (ApplicationId appId : deployed) {
+        try {
+          helper.deleteAppVersion(appId);
```

Comment:
If we're deleting, we shouldn't do it one by one. That being said, I don't think we need to (or should) delete, as there is no concept of deleting just a single app version today.

Reply:
We do have `appLifecycleService.removeApplicationVersion` which I was referring ton but looks like the public API is deprecated. 
The main concern is without deletion this versions left would be surfaced to the user in history creating confusion. Specially in case of a pull of new application as the application would have only one version which is nor marked latest.

Reply:
this isn't the right way to prevent these versions from surfacing in the history for a couple reasons:

1. It's always possible to see it after the app is added, but before it is marked as latest
2. The delete can fail

If we are concerned about showing it, the right way to handle it is to add something like an isHidden flag. Trying to delete it here is like a half solution, so I don't think we need to bother doing it.

Reply:
I agree. Rather than adding a additional flag I think we can have a enum based state field which can be extended later and covers both `latest` status and `hidden` status. I have added a followup Jira for this as this is not a blocker for the core feature.